File: app/src/main/java/deltazero/amarok/ui/PasswordAuthFragment.java
package deltazero.amarok.ui;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

import deltazero.amarok.PrefMgr;
import deltazero.amarok.R;
import deltazero.amarok.utils.HashUtil;

public class PasswordAuthFragment extends BottomSheetDialogFragment {

    private OnVerifiedCallback onVerifiedCallback;
    private TextInputEditText etPassword;
    private TextInputLayout tilPassword;

    public interface OnVerifiedCallback {
        void onVerified(boolean succeed);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        setCancelable(false);
        View fragmentView = inflater.inflate(R.layout.dialog_security, container, false);

        fragmentView.findViewById(R.id.security_dialog_bt_unlock).setOnClickListener(v -> verify());
        fragmentView.findViewById(R.id.security_dialog_bt_cancel).setOnClickListener(v -> {
            if (onVerifiedCallback != null) onVerifiedCallback.onVerified(false);
        });
        etPassword = fragmentView.findViewById(R.id.security_dialog_et_password_input);
        tilPassword = fragmentView.findViewById(R.id.security_dialog_til_password_input);

        // Clear error on text changed
        etPassword.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                tilPassword.setError(null);
                tilPassword.setErrorEnabled(false);
            }
        });

        return fragmentView;
    }

    public PasswordAuthFragment setOnVerifiedCallback(OnVerifiedCallback onVerifiedCallback) {
        this.onVerifiedCallback = onVerifiedCallback;
        return this;
    }

    private void verify() {

        String password = PrefMgr.getAmarokPassword();
        assert etPassword.getText() != null;

        if (password == null || HashUtil.calculateHash(etPassword.getText().toString()).equals(password)) {
            if (onVerifiedCallback != null) onVerifiedCallback.onVerified(true);
            dismiss();
        } else {
            tilPassword.setError(getText(R.string.password_incorrect));
        }
    }

}

File: app/src/main/java/deltazero/amarok/ui/PasswordAuthFragment.kt
package deltazero.amarok.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.material3.Button
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import dagger.hilt.android.AndroidEntryPoint
import deltazero.amarok.R
import deltazero.amarok.core.SettingsRepository
import deltazero.amarok.ui.theme.AmarokTheme
import deltazero.amarok.utils.HashUtil
import javax.inject.Inject

@AndroidEntryPoint
class PasswordAuthFragment : BottomSheetDialogFragment() {
  @Inject lateinit var settingsRepo: SettingsRepository

  private var onVerifiedCallback: OnVerifiedCallback? = null

  fun interface OnVerifiedCallback {
    fun onVerified(succeed: Boolean)
  }

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?,
  ): View {
    isCancelable = false
    return ComposeView(requireContext()).apply {
      setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
      setContent {
        AmarokTheme {
          PasswordAuthSheet(
            isCorrect = { input ->
              val password = settingsRepo.settings.value.password
              password == null || HashUtil.calculateHash(input) == password
            },
            onCancel = { onVerifiedCallback?.onVerified(false) },
            onVerified = {
              onVerifiedCallback?.onVerified(true)
              dismiss()
            },
          )
        }
      }
    }
  }

  fun setOnVerifiedCallback(onVerifiedCallback: OnVerifiedCallback?): PasswordAuthFragment {
    this.onVerifiedCallback = onVerifiedCallback
    return this
  }
}

@Composable
private fun PasswordAuthSheet(
  isCorrect: (String) -> Boolean,
  onCancel: () -> Unit,
  onVerified: () -> Unit,
) {
  var password by remember { mutableStateOf("") }
  var showError by remember { mutableStateOf(false) }

  Column(
    modifier =
      Modifier.fillMaxWidth().padding(start = 40.dp, end = 40.dp, top = 40.dp, bottom = 50.dp)
  ) {
    Row(verticalAlignment = Alignment.CenterVertically) {
      Icon(Icons.Outlined.Lock, contentDescription = null, modifier = Modifier.size(20.dp))
      Spacer(Modifier.width(5.dp))
      Text(stringResource(R.string.unlock_required), style = MaterialTheme.typography.titleMedium)
    }
    OutlinedTextField(
      value = password,
      onValueChange = {
        password = it
        showError = false
      },
      modifier = Modifier.fillMaxWidth().padding(top = 20.dp),
      label = { Text(stringResource(R.string.password)) },
      singleLine = true,
      isError = showError,
      supportingText =
        if (showError) {
          { Text(stringResource(R.string.password_incorrect)) }
        } else null,
      visualTransformation = PasswordVisualTransformation(),
      keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.NumberPassword),
    )
    Row(
      modifier = Modifier.fillMaxWidth().padding(top = 15.dp),
      horizontalArrangement = Arrangement.End,
    ) {
      TextButton(onClick = onCancel) { Text(stringResource(R.string.cancel)) }
      Button(onClick = { if (isCorrect(password)) onVerified() else showError = true }) {
        Text(stringResource(R.string.unlock))
      }
    }
  }
}
